Reader: only show following intro banner to users < 2 weeks old

bluefuton opened the pull request:
The Reader following intro banner is currently shown to all users until they dismiss it or use the Manage Following link within it.

This PR autohides the banner if the user has been registered for more than two weeks.

<img src="https://user-images.githubusercontent.com/17325/27074684-bc56d26c-501f-11e7-878b-05c5b26c568c.png" />

Fixes #14824.

Pre-requisite: https://github.com/Automattic/wp-calypso/pull/15029

blowery commented on client/state/ui/guided-tours/contexts.js:
```diff
@@ -62,8 +62,18 @@ const timeSinceUserRegistration = state => {
  * @return {Boolean} True if user is new, false otherwise
  */
 export const isNewUser = state => {
+	return isUserNewerThan( state, WEEK_IN_MILLISECONDS );
```
interesting, no test coverage in `./test/contexts`. maybe we should add some?

bluefuton replied:
I had the same thought, but not sure how to mock `Date.now()` in `timeSinceUserRegistration`?

blowery replied:
ah, http://sinonjs.org/releases/v2.3.4/fake-timers/

bluefuton replied:
Used `moment` and added some basic tests for `isUserNewerThan`.

blowery commented on client/reader/following/intro.jsx:
```diff
@@ -15,6 +15,7 @@ import { savePreference } from 'state/preferences/actions';
 import { getCurrentUserId } from 'state/current-user/selectors';
+import { isUserNewerThan, WEEK_IN_MILLISECONDS } from 'state/ui/guided-tours/contexts';
 
 const isEven = number => number % 2 === 0;
```
can drop this now

blowery replied:
ah, cool!

blowery commented on client/reader/following/intro.jsx:
```diff
@@ -42,11 +43,11 @@ class FollowingIntro extends React.Component {
+		const { isNewReader, translate, dismiss, userId, isNewUser } = this.props;
 		const linkElement = <a onClick={ this.props.handleManageLinkClick } href="/following/manage" />;
 
 		// Only show the banner to new Readers with an odd user ID (simple A/B test)
```
update the comment

bluefuton replied:
Also #15029

blowery replied:
guessing this gets dropped / updated in #15029 too

bluefuton commented on client/state/ui/guided-tours/test/contexts.js:
```diff
@@ -14,7 +15,10 @@ import {
 	SOURCE_UNKNOWN,
 } from 'components/tinymce/plugins/wpcom-track-paste/sources';
 
+const WEEK_IN_MILLISECONDS = 7 * 1000 * 3600 * 24
```
Duplicating this const in the test because including it like this:

`import { WEEK_IN_MILLISECONDS } from 'state/ui/guided-tours/contexts';`

...causes it to bail, presumably because of what else contexts.js imports.

bluefuton replied:
Could potentially include it inside the `useMockery` block.

blowery replied:
yup, though it'll have to be a require instead of an import.

samouri commented:
I don't think this worked for me.  I tested out an account that is months old (but would get the intro banner before this PR) and it still has the intro banner :(

blowery commented:
>  I tested out an account that is months old (but would get the intro banner before this PR) and it still has the intro banner :(.

Set a break in the debugger and see what's busted?

blowery commented:
good catch @samouri ! pushed a fix in https://github.com/Automattic/wp-calypso/pull/15052/commits/84ac2e52a472a20b44fc6300f1b2a6461e2a3f98

tricky curried functions.

blowery commented:
took me a bit to track this down. is_new_reader is a calypso pref, which isn't what the wp-cli script sets (via `wp user-attributes mark_is_new_reader`)

bluefuton commented:
Well spotted @samouri, and thank for the fix @blowery! We were talking about curried functions whilst I was working on this, and I updated the tests but not the main usage in intro.jsx.

`is_new_reader` used to be a user attribute, but we switched to Calypso prefs so we could take advantage of Redux goodness to get and set the preference.

blowery commented:
@samouri can you try this one more time with your old new user? :)
